python/polars_ds/num.py

```diff
@@ -1066,3 +1066,42 @@ def _haversine(
             is_elementwise=True,
             cast_to_supertypes=True,
         )
+
+    def knn_entropy(self, *others: pl.Expr, k=2):
+        """
+        Multivariate estimation of entropy through a k-nearest neighbor query
+
+        Reference
+        ---------
+        https://arxiv.org/pdf/1506.06501v1.pdf
+        """
+        from scipy.special import digamma
+        from math import pi, gamma, log
+        (N, d) = pl.len(), len(others) + 1  # d has to include self
+        g1 = digamma(N) - digamma(k)
+        index = pl.int_range(N, dtype=pl.UInt32).alias("_idx")
+        neighbor = index.num.knn_ptwise(self._expr, *others, dist='l1', k=k).list.get(k)
+        # euclidean distance computation for now. But others can be used
+        dists = pl.sum_horizontal(pl.exclude('_idx').sub(pl.exclude('_idx').gather(neighbor)).pow(2)).sqrt()
+        logd = dists.mul(2).log().mul(d).truediv(N).sum()
+        cd = pi ** (d / 2) / 2 ** d / gamma(1 + d / 2)
+        return (g1 + log(cd) + logd)
+
+
+    def copula_entropy(self, *others, k=2):
```

Comment:
Please add types to all funtions and eventually add some comments and references when you are done. Thanks